src/zenml/models/v2/base/base.py

```diff
@@ -0,0 +1,225 @@
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Base model definitions."""
+
+from datetime import datetime
+from typing import Any, ClassVar, Dict, Generic, List, Optional, TypeVar
+from uuid import UUID
+
+from pydantic import Field, SecretStr
+from pydantic.generics import GenericModel
+
+from zenml.exceptions import HydrationError
+from zenml.utils.pydantic_utils import YAMLSerializationMixin
+
+# -------------------- Base Model --------------------
+
+
+class BaseZenModel(YAMLSerializationMixin):
+    """Base model class for all ZenML models.
+
+    This class is used as a base class for all ZenML models. It provides
+    functionality for tracking analytics events and proper encoding of
+    SecretStr values.
+    """
+
+    ANALYTICS_FIELDS: ClassVar[List[str]] = []
+
+    def get_analytics_metadata(self) -> Dict[str, Any]:
+        """Get the analytics metadata for the model.
+
+        Returns:
+            Dict of analytics metadata.
+        """
+        metadata = {}
+        for field_name in self.ANALYTICS_FIELDS:
+            metadata[field_name] = getattr(self, field_name, None)
+        return metadata
+
+    class Config:
+        """Pydantic configuration class."""
+
+        # This is needed to allow the REST client and server to unpack SecretStr
+        # values correctly.
+        json_encoders = {
+            SecretStr: lambda v: v.get_secret_value()
+            if v is not None
+            else None
+        }
+
+        # Allow extras on all models to support forwards and backwards
+        # compatibility (e.g. new fields in newer versions of ZenML servers
+        # are allowed to be present in older versions of ZenML clients and
+        # vice versa).
+        extra = "allow"
+
+
+# -------------------- Request Model --------------------
+
+
+class BaseRequest(BaseZenModel):
+    """Base request model.
+
+    Used as a base class for all request models.
+    """
+
+
+# -------------------- Response Model --------------------
+
+
+class BaseResponseBody(BaseZenModel):
+    """Base body model.
+
+    Used as a base class for all body models associated with responses.
+    Features a creation and update timestamp.
+    """
+
+    created: Optional[datetime] = Field(
+        title="The timestamp when this resource was created.",
+        default=None,
+    )
+    updated: Optional[datetime] = Field(
+        title="The timestamp when this resource was last updated.",
+        default=None,
+    )
+
+
+class BaseResponseMetadata(BaseZenModel):
+    """Base metadata model.
+
+    Used as a base class for all metadata models associated with responses.
+    """
+
+
+AnyBody = TypeVar("AnyBody", bound=BaseResponseBody)
+AnyMetadata = TypeVar("AnyMetadata", bound=BaseResponseMetadata)
```

Comment:
The `BaseResponse` is a generic model class. It uses the provided types to adjust the signature of its methods such as `get_hydrated_version`, `_validate_hydrated_version`, and its respective `body` and `metadata` fields. This is very significant in `mypy` terms as well as it helps a lot with lining. When using the `Generic[..., ...]` we needed the `AnyBody` and `AnyMetadata`, otherwise, if we use the class definitions directly we end up with the error: `Parameters to 'Generic[...]' must all be type variables `.